### ISSUE TITLE
Add a way to convert tokens for scalars in PHP code to PHP equivalents (int/float/string/bool/null literals)

### ISSUE THREAD

**Issue by TysonAndre** [Microsoft/tolerant-php-parser#180]
(For the usability phase)

Motivations:

- Language server may need to extract values from the AST for conditionally emitting warnings
- API doesn't provide a way to convert a token fragment for a literal to the scalar it represents
- Need to trim strings and numbers manually in order to extract them, may not notice edge cases
- Various edge cases that would be time consuming to re-implement in projects using the parser, such as converting the tokens used to represent parts of `   "Multi $varName part"` into individual parts without losing this `"Multi "`
  
   And the various forms of heredoc

   And removing backslashes, escaping
- There are multiple forms of integers, those forms may change in future releases

Possible API:

`Token->representsScalarLiteral(): bool` (Checks TokenKind)
`Token->asPHPScalarForLiteral() : int/float/string/bool/null` (Throw exception if representsScalarLiteral() would be false.

**Comment by roblourens** [comment]
Yes, this would be useful. Are you using tolerant-php-parser for a project currently? Evaluating it?

**Comment by TysonAndre** [comment]
I was considering using that. Currently, I'm using https://github.com/TysonAndre/php-parser-to-php-ast (Based on nikic/php-parser) for https://github.com/phan/phan/pull/1110



https://github.com/TysonAndre/tolerant-php-parser-to-php-ast is a broken work in progress, converting from one AST to another involves a lot of time. (Haven'tt fixed parts copied from my other project)

**Comment by roblourens** [comment]
Cool. Parsing invalid PHP is an explicit design goal of this project, so https://github.com/phan/phan/pull/1110 sounds like a good usecase for it.

There are some parts which are not fully fleshed out, like the issues you've opened. I'd be happy to support you if something is blocking you or if you want to work on a PR for one of these issues.

**Comment by TysonAndre** [comment]
Major blockers include:

- extracting scalars(could probably hack together the common case with filter_val(FILTER_VALIDATE_INT) and unescaping backslashes(e.g. use php-parser's code for that)
- extracting doc comments for methods, closures, etc. even with things in between (probably doable by walking the tree)
- grammar bugs - invalid statements like `2 = $x` may generate nodes that cause unexpected exceptions or errors

Other issues won't be known until this is close to done.

**Comment by TysonAndre** [comment]
See https://github.com/TysonAndre/tolerant-php-parser-to-php-ast/blob/53e2b53707e98090be9430332b7929054f9bf902/src/TolerantASTConverter/TolerantASTConverter.php#L835 for an attempt at this, currently relies on php-parser's implementation.
